src/libxively/cbor/xi_cbor_codec_ct.c

```diff
@@ -0,0 +1,344 @@
+/* Copyright (c) 2003-2017, LogMeIn, Inc. All rights reserved.
+ *
+ * This is part of the Xively C Client library,
+ * it is licensed under the BSD 3-Clause license.
+ */
+
+#include <xi_cbor_codec_ct.h>
+
+#include <cbor.h>
+#include <cn-cbor/cn-cbor.h>
+
+#include <xively_error.h>
+#include <xi_macros.h>
+
+#include <xi_debug.h>
+
+#ifdef USE_CBOR_CONTEXT
+
+#include <xi_allocator.h>
+
+void* xi_cn_calloc_func_wrapper( size_t count, size_t size, void* context )
+{
+    ( void )context;
+
+    // printf( "hello calloc wrapper: %lu, %lu\n", count, size );
+
+    return xi_calloc( count, size );
+}
+
+void xi_cn_free_func_wrapper( void* ptr, void* context )
+{
+    ( void )context;
+
+    xi_free( ptr );
+}
+
+static cn_cbor_context cn_cbor_context_object = {.calloc_func = xi_cn_calloc_func_wrapper,
+                                                 .free_func   = xi_cn_free_func_wrapper,
+                                                 .context     = NULL};
+
+cn_cbor_context* context = &cn_cbor_context_object;
+
+#endif
+
+
+#define XI_CBOR_CODEC_CT_STRING_MSGTYPE "msgtype"
+#define XI_CBOR_CODEC_CT_STRING_MSGVER "msgver"
+#define XI_CBOR_CODEC_CT_STRING_LIST "list"
+#define XI_CBOR_CODEC_CT_STRING_FILE_NAME "N"
+#define XI_CBOR_CODEC_CT_STRING_FILE_REVISION "R"
+#define XI_CBOR_CODEC_CT_STRING_FILE_OPERATION "O"
+#define XI_CBOR_CODEC_CT_STRING_FILE_IMAGESIZE "S"
+#define XI_CBOR_CODEC_CT_STRING_FILE_FINGERPRINT "F"
+
+#define XI_CBOR_CODEC_CT_STRING_FILECHUNK_OFFSET "O"
+#define XI_CBOR_CODEC_CT_STRING_FILECHUNK_LENGTH "L"
+
+void xi_cbor_put_name_and_revision( cn_cbor* cb_map,
+                                    const char* name,
+                                    const char* revision,
+                                    cn_cbor_errback* errp )
+{
```

Comment:
This is an internally used function and NULL check is done later. I think quality is enough here. In general on this level we may allow messages without name and/or revision. Although is isn't used without name and revision. So I'd leave this as it is for now if it is ok for you.